### PR TITLE
Revoke associated JWTs on user deletion

### PR DESCRIPTION
**Pull request by scottinet**
# Description

When deleting a user using `security:deleteUser`, the associated JWT were kept for their entire duration.
This lead to bugs when a deleted user tried to use one of these tokens, as Kuzzle tried to load a non-existing user, trusting the token to know if such a user exist.

This pull request maintain a list of JWT per user, and deletes all associated JWT when a user gets deleted.

Functional tests have been updated to reproduce the problem, ensuring it has been resolved.

# Other changes

* solve sonarqube issues on modified files
* rewrite most of the `authController` unit tests file, as it wasn't using mocks and because many tests  succeeded but not on the expected behavior
* fix unit tests on the `TokenRepository` class

# Solved issue

#773

### CODE REVIEW

File: lib/api/controllers/securityController.js [hunk 1]
```diff
@@ -389,6 +412,7 @@ class SecurityController {
     assertHasId(request);
 
     return this.kuzzle.repositories.user.delete(request.input.resource._id)
+      .then(() => this.kuzzle.repositories.token.deleteByUserId(request.input.resource._id))
```
Comment (scottinet) [inline lib/api/controllers/securityController.js]: fix: now clean up JWTs after the user has been successfully deleted

File: lib/api/core/models/repositories/tokenRepository.js [hunk 2]
```diff
@@ -1,47 +1,68 @@
...
   }
 
   /**
    * @param {Token} requestToken
    * @returns {Promise<Object>}
    */
-  expire (requestToken) {
+  expire(requestToken) {
     return super.expireFromCache(requestToken)
+      .then(() => super.delete(requestToken.userId + '#' + requestToken._id))
```
Comment (scottinet) [inline lib/api/core/models/repositories/tokenRepository.js]: Clean up userid<=>token link on a forced token expiration

File: lib/api/core/models/repositories/tokenRepository.js [hunk 3]
```diff
@@ -100,21 +121,22 @@ class TokenRepository extends Repository {
       .then(() => {
         this.kuzzle.tokenManager.add(token, request.context);
 
-        return token;
+        return this.persistToCache({_id: user._id + '#' + token._id}, expiresIn / 1000);
```
Comment (scottinet) [inline lib/api/core/models/repositories/tokenRepository.js]: Add a userid<=>token link on a token creation

File: lib/api/core/models/repositories/tokenRepository.js [hunk 4]
```diff
@@ -189,25 +211,62 @@ class TokenRepository extends Repository {
...
     return this.serializeToCache (token);
   }
+
+  /**
+   * Deletes tokens affiliated to the provided user identifier
+   *
+   * @param {string} userId
+   * @return {Promise}
+   */
+  deleteByUserId(userId) {
```
Comment (scottinet) [inline lib/api/core/models/repositories/tokenRepository.js]: New method: delete all existing tokens created by a user ID

File: lib/api/controllers/securityController.js [hunk 5]
```diff
@@ -1,23 +1,48 @@
+/*
```
Comment (ballinette) [inline lib/api/controllers/securityController.js]: I don't know if the licence preamble is mandatory in the beginning of all the source files, or if the `LICENCE.md` file at the project's root is enough.
But if this preamble is mandatory here, it should be in ALL the files (should be done in a dedicated PR, thus)

File: lib/api/core/auth/tokenManager.js [hunk 6]
```diff
@@ -22,38 +43,43 @@ const TIMEOUT_MAX = Math.pow(2, 31) - 1;
...
     if (this.tokens.array.length > 0) {
-      let delay = Math.min(this.tokens.array[0].expiresAt - Date.now(), TIMEOUT_MAX);
+      const delay = Math.min(this.tokens.array[0].expiresAt - Date.now(), TIMEOUT_MAX);
 
       if (this.timer) {
         clearTimeout(this.timer);
       }
 
-      this.timer = setTimeout(this.checkTokensValidity, delay);
+      this.timer = setTimeout(this.checkTokensValidity.bind(this), delay);
```
Comment (ballinette) [inline lib/api/core/auth/tokenManager.js]: Why do we need `bind(this)` ?

Comment (scottinet) [inline lib/api/core/auth/tokenManager.js]: because otherwise the method is run with an empty context

File: lib/api/controllers/securityController.js (hunk 5 above)
Comment (scottinet) [inline lib/api/controllers/securityController.js]: It is required by the Apache 2.0 license. And a PR is already underway, I'm updating files in this PR to ease up future merges.

File: lib/api/core/auth/tokenManager.js (hunk 6 above)
Comment (ballinette) [inline lib/api/core/auth/tokenManager.js]: Oh yes, I did not notice the setTimeout !! (yes, my glasses may be dirty... :/ )

Comment (ballinette) [inline lib/api/core/auth/tokenManager.js]: Oh yes, I did not notice the `setTimeout` !! (probable my glasses are dirty... :/ )
